controller/model/controller_setting_manager.go

```diff
@@ -0,0 +1,204 @@
+	baseEntityManager[*ControllerSetting, *db.ControllerSetting]
+}
+
+// ReadEffective returns a controller setting object that contains the effective settings and instance settings for a
+// specific controller. Effective settings are global + instance overrides.
+func (s *ControllerSettingManager) ReadEffective(id string) (*ControllerSettingEffective, error) {
+
+	var global, instance, effective *ControllerSetting
+
+	err := s.env.GetDb().View(func(tx *bbolt.Tx) error {
+		var err error
+		globalSt, err := s.Store.LoadById(tx, db.ControllerSettingGlobalId)
+
+		if err != nil {
+			return fmt.Errorf("could not read global settings: %w", err)
+		}
+
+		global = &ControllerSetting{}
+
+		err = global.fillFrom(s.env, tx, globalSt)
+
+		if err != nil {
+			return fmt.Errorf("could not fill global settings: %w", err)
+		}
+
+		// if reading effective of global, don't do any more work, global == effective == instance
+		if id == db.ControllerSettingGlobalId {
+			instance = global
+			effective = global
+			return nil
+		}
+
+		instanceSt, err := s.Store.LoadById(tx, id)
+
+		if err != nil && !boltz.IsErrNotFoundErr(err) {
+			return fmt.Errorf("could not read instance settings: %w", err)
+		}
+
+		if instanceSt != nil {
+			instance = &ControllerSetting{}
+			err = instance.fillFrom(s.env, tx, instanceSt)
+
+			effectiveSt := instanceSt.MergeSettings(globalSt)
+
+			err = effective.fillFrom(s.env, tx, effectiveSt)
+
+			if err != nil {
+				return fmt.Errorf("could not fill effective settings: %w", err)
+			}
+		} else {
+			effective = global
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &ControllerSettingEffective{
+		Effective: effective,
+		Instance:  instance,
+	}, nil
+}
+
+func (s *ControllerSettingManager) newModelEntity() *ControllerSetting {
+	return &ControllerSetting{
+		BaseEntity: models.BaseEntity{},
+		Oidc: &OidcSettings{
+			OidcSettingDef: &db.OidcSettingDef{
+				RedirectUris:   nil,
+				PostLogoutUris: nil,
+			},
+		},
+	}
+}
+
+type ControllerSettingListener func(settingPath string, setting *ControllerSetting)
+
+func (s *ControllerSettingManager) Watch(settingPath string, listener func(string, *ControllerSetting)) {
+	s.env.GetStores().ControllerSetting.Watch(settingPath, func(setting string, controllerId string, settingEvent *db.ControllerSettingsEvent) {
+
+		controllerSettingModel := &ControllerSetting{
+			BaseEntity: models.BaseEntity{},
+			Oidc: &OidcSettings{
+				OidcSettingDef: &db.OidcSettingDef{},
+			},
+		}
+
+		controllerSettingModel.BaseEntity.FillCommon(settingEvent.Effective)
+		controllerSettingModel.Oidc.OidcSettingDef = settingEvent.Effective.Oidc
+
+		listener(settingPath, controllerSettingModel)
+	}, s.env.GetControllerId())
+}
+
+func (s *ControllerSettingManager) Marshall(entity *ControllerSetting) ([]byte, error) {
+	jsonData, err := json.Marshal(entity)
+	if err != nil {
+		return nil, fmt.Errorf("could not marshal entity: %w", err)
+	}
+
```

Comment:
I'm doing some json in protobuf as well for inspect stuff. I'll look at this and see if it makes sense to use there.